evm/vm/base.py

```diff
@@ -484,7 +491,31 @@ def finalize_block(self, block):
         # TODO: Refactor to only use batching/journaling for tx processing
         self.state.account_db.persist()
 
-        return block.copy(header=block.header.copy(state_root=self.state.state_root))
+        pre_mining_header = block.header.copy(state_root=self.state.state_root)
+        cache = mkcache_bytes(block.number)
+        difficulty = block.header.difficulty
+        for i in range(100000):
+            nonce = os.urandom(8)
+            mining_output = hashimoto_light(
+                block.number, cache, pre_mining_header.mining_hash,
+                big_endian_to_int(nonce))
+            result = big_endian_to_int(mining_output[b'result'])
+            try:
+                validate_lte(result, 2**256 // difficulty, title="POW Difficulty")
+            except ValidationError:
+                pass
+            else:
+                print("==============================")
+                print("Found mix_hash after %d attempts" % i)
+                print("==============================")
+                break
+        else:
+            raise Exception("Too many attempts, giving up")
+        header = pre_mining_header.copy(
+            nonce=nonce,
+            mix_hash=mining_output[b'mix digest'])
```

Comment:
Looks like the original question is mostly resolved. The first thing I noticed is that this code block can be extracted with a pretty nice interface of `(block.number. mining_hash) -> (nonce, mix_hash)`.  That API only makes sense for "test mining" though since real mining is going to be a simultaneous mix of constantly maximizing gas in the block while searching for a nonce. Which means that managing "real mining" probably belongs out in `trinity` anyway.